File: N-Recycler-Lambda-Example/src/main/java/tech/thdev/recycler_lambda_example/base/BaseRecyclerViewHolder.java
package tech.thdev.recycler_lambda_example.base;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

/**
 * Created by Tae-hwan on 3/28/16.
 */
public abstract class BaseRecyclerViewHolder<AD extends BaseRecyclerAdapter, DATA extends AdapterItem> extends RecyclerView.ViewHolder {

    private AD adapter;

    public BaseRecyclerViewHolder(AD adapter, View itemView) {
        super(itemView);

        this.adapter = adapter;
    }

    public BaseRecyclerViewHolder(@LayoutRes int layoutResId, ViewGroup parent, AD adapter) {
        this(adapter, LayoutInflater.from(adapter.getContext()).inflate(layoutResId, parent, false));
    }

    public abstract void onViewHolder(final @Nullable DATA item, int position);

    protected AD getAdapter() {
        return adapter;
    }

    protected Context getContext() {
        return adapter.getContext();
    }
}

File: Android/2016-07-21-kotlin-example-01/app/src/main/kotlin/tech/thdev/kotlin_example_01/base/adapter/BaseRecyclerViewHolder.kt
package tech.thdev.kotlin_example_01.base.adapter

import android.content.Context
import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import tech.thdev.kotlin_example_01.base.model.BaseItem

/**
 * Created by Tae-hwan on 7/21/16.
 */
abstract class BaseRecyclerViewHolder<ITEM: BaseItem>(protected val adapter: BaseRecyclerAdapter<ITEM>, itemView: View) : RecyclerView.ViewHolder(itemView) {

    constructor(@LayoutRes layoutRes: Int, parent: ViewGroup?, adapter: BaseRecyclerAdapter<ITEM>)
        : this(adapter, LayoutInflater.from(adapter.context).inflate(layoutRes, parent, false)) {
    }

    abstract fun onViewHolder(item: ITEM?, position: Int)

    protected val context: Context
        get() = adapter.context
}